### PR TITLE
[STRATCONN-2144] Make error code and status mandatory for Integration Error

### PR DESCRIPTION
**Pull request by varadarajan-tw**
This PR marks error code and status as mandatory for Integration Error class. Error code is important because they are shown in Event Delivery and Delivery Overview screens and indicates the type of error that has occurred. Status is influential in retries. 
Marking them mandatory would enforce builders to spend time providing appropriate error codes and status thereby improving observability.

Errors without error code are tagged as `Internal` errors and show up in Sentry. This happened for [Heap](https://twilio.slack.com/archives/C04J8QJ1Y59/p1679078393576379) and hence we are taking this effort. 

Destinations that weren't providing error codes or status codes have been cleaned up and deployed already as part of different PRs.

## Testing

All existing unit tests are passing. 

- [ ] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [ ] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

### CODE REVIEW

File: packages/destination-actions/src/destinations/fullstory/__tests__/fullstory.test.ts [hunk 1]
```diff
@@ -162,7 +162,7 @@ describe('FullStory', () => {
     falsyUserIds.forEach((falsyUserId) => {
       it(`it throws PayloadValidationError given falsy user id ${falsyUserId}`, async () => {
         await expect(testDestination.onDelete!({ type: 'delete', userId: falsyUserId }, settings)).rejects.toThrowError(
-          new IntegrationError('User Id is required for user deletion.')
+          new PayloadValidationError('User Id is required for user deletion.')
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/fullstory/__tests__/fullstory.test.ts:165]: I missed updating this test as part of https://github.com/segmentio/action-destinations/pull/1299

**Review by joe-ayoub-segment — APPROVED** [review]
Hi @varadarajan-tw , 
The code change looks simple enough. test pass. yarn validate passes. 

What did you do to ensure that all the references to IntegrationError provide all required params? I looked through about 50 references and they looked good. 

The change to Criteo's error class looks safe.

**Comment by varadarajan-tw** [comment]
@joe-ayoub-segment  - Over the past 2-3 weeks, I fixed Integration errors that didn't have code or status code one by one and merged them separately. They have all been deployed to production now and no issues have been observed so far. This final PR is to mark code and status code mandatory so that Builders provide a code and status ensuring that we have sufficient information in the Event Delivery/ Delivery Overview screens.

https://github.com/segmentio/action-destinations/pull/1321
https://github.com/segmentio/action-destinations/pull/1310
https://github.com/segmentio/action-destinations/pull/1309
https://github.com/segmentio/action-destinations/pull/1296
https://github.com/segmentio/action-destinations/pull/1299
https://github.com/segmentio/action-destinations/pull/1295

**Review by joe-ayoub-segment — APPROVED** [review]

**Review by SyedWasiHaider — APPROVED** [review]

File: packages/core/src/errors.ts [hunk 2]
```diff
@@ -12,10 +12,12 @@ export class IntegrationError extends CustomError {
...
    * @param message - a human-friendly message to display to users
-   * @param code - an optional error code/reason
-   * @param status - an optional http status code (e.g. 400)
+   * @param code - error code/reason
+   * @param status - http status code (e.g. 400).
+   *    - 4xx errors are not automatically retried, except for 408, 423, 429
+   *    - 5xx are automatically retried, except for 501
    */
-  constructor(message = '', code?: string, status?: number) {
+  constructor(message: string, code: string, status: number) {
```
[the remark this answers is not in the document]
Comment (varadarajan-tw) [reply packages/core/src/errors.ts:20]: We can restrict code to be of ErrorCode enum type. That would require another round of clean up. Most of them pass a string as error code.

**Pull request merged**